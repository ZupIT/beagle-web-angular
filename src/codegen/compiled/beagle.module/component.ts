/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import { Type } from '@angular/core'
import {
  kebabToCamelCase,
  removeExtraIndentation,
  replaceToUnderline,
} from '../../utils/formatting'
import { getComponentAnnotations } from '../../utils/metadata'
import { remoteViewSelector } from '../../../constants'


function createViewChildString(name: string, templateName: string, angularVersion: number): string {
  const componentName = replaceToUnderline(name)

  if (angularVersion >= 8) {
    return `'${componentName}': new ViewChild('${templateName}', { static: true })`
  } else {
    return `'${componentName}': new ViewChild('${templateName}')`
  }
}

function createQueries(components: Record<string, Type<any>>, angularVersion: number) {
  const componentNames = Object.keys(components)

  return componentNames.map((name) => {
    const selector = getComponentAnnotations(components[name]).selector
    const templateName = kebabToCamelCase(selector)
    return createViewChildString(name, templateName, angularVersion)
  })
}


export function createComponentString(
    components: Record<string, Type<any>>,
    angularVersion: number,
  ) {
  const queries = createQueries(components, angularVersion)

  // todo: legacy code. Remove the input "loadParams" with v2.0.
  const componentString = `
    @Component({
      selector: '${remoteViewSelector}',
      template,
      inputs: ['loadParams', 'route', 'networkOptions', 'controllerId'],
      queries: {
        ${queries.join(',\n        ')},
      },
    })
    export class BeagleRemoteView extends AbstractBeagleRemoteView {
      constructor(
        beagleProvider: BeagleProvider,
        ngZone: NgZone,
        changeDetector: ChangeDetectorRef,
      ) {
        // @ts-ignore
        super(beagleProvider, ngZone, changeDetector)
      }
    }
  `

  return removeExtraIndentation(componentString, 4)
}
