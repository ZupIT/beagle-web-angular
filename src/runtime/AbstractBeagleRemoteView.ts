/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import {
  AfterViewInit,
  OnDestroy,
  OnChanges,
  SimpleChanges,
  ChangeDetectorRef,
  NgZone,
  TemplateRef,
  Output,
  EventEmitter,
} from '@angular/core'
import {
  LoadParams,
  BeagleView,
  IdentifiableBeagleUIElement,
  replaceBindings,
  EventHandler,
  createEventHandler,
} from '@zup-it/beagle-web'
import { BeagleContext } from '@zup-it/beagle-web'
import { twoPointsToUnderline } from '../codegen/utils/formatting'
import { AbstractBeagleProvider } from './AbstractBeagleProvider'
import { createStaticPromise } from './utils/promise'
import BeagleRuntimeError from './errors'
import beagleMapKeysConfig from './utils/beagle-map-keys-config'

let nextViewId = 1

export abstract class AbstractBeagleRemoteView implements AfterViewInit, OnDestroy, OnChanges {
  loadParams: LoadParams = { path: '' }
  tree: IdentifiableBeagleUIElement<any>
  view: BeagleView
  viewId = `${nextViewId++}`
  beagleProvider: AbstractBeagleProvider
  ngZone: NgZone
  changeDetector: ChangeDetectorRef
  viewStaticPromise = createStaticPromise<BeagleView>()
  eventHandler: EventHandler

  @Output() onCreateBeagleView: EventEmitter<BeagleView> = new EventEmitter<BeagleView>()

  constructor(
    beagleProvider?: AbstractBeagleProvider,
    ngZone?: NgZone,
    changeDetector?: ChangeDetectorRef,
  ) {
    if (beagleProvider) this.beagleProvider = beagleProvider
    if (ngZone) this.ngZone = ngZone
    if (changeDetector) this.changeDetector = changeDetector
  }

  createBeagleView() {
    const beagleService = this.beagleProvider.getBeagleUIService()
    if (!beagleService) {
      throw new BeagleRuntimeError(
        'you need to start the beagle provider before using a remote view.',
      )
    }
    this.view = beagleService.createView(this.loadParams.path)
    this.view.subscribe(this.updateView)
    this.view.addErrorListener((errorListener) => {
      errorListener.forEach((error) => {
        console.error(error)
      })
    })
    BeagleContext.registerView(`${this.viewId}`, this.view)
    this.viewStaticPromise.resolve(this.view)
    this.onCreateBeagleView.emit(this.view)
  }

  createEventHandler() {
    const beagleService = this.beagleProvider.getBeagleUIService()
    if (!beagleService) return
    this.eventHandler = createEventHandler(beagleService.getConfig().customActions, this.view)
  }

  getTemplate(componentName: IdentifiableBeagleUIElement<any>['type']): TemplateRef<any> {
    const component = beagleMapKeysConfig.getComponent(componentName)
    const normalizedComponentName = twoPointsToUnderline(component)
    
    if (!this[normalizedComponentName]) {
      console.warn(
        `Beagle: the component ${componentName} was not declared in Beagle's configuration.`,
      )
    }
    return this[normalizedComponentName]
  }

  updateView = (uiTree: IdentifiableBeagleUIElement<any>) => {
    this.ngZone.run(() => {
      const uiTreeWithActions = this.eventHandler.interpretEventsInTree(uiTree)
      const uiTreeWithValues = replaceBindings(uiTreeWithActions)
      this.tree = uiTree && Object.keys(uiTree).length > 0 ? uiTreeWithValues : null
      this.changeDetector.detectChanges()
    })
  }

  elementIdentity(index: number, element: IdentifiableBeagleUIElement<any>) {
    return element.id
  }

  getView() {
    return this.viewStaticPromise.promise
  }

  ngAfterViewInit() {
    if (!this.beagleProvider || !this.ngZone || !this.changeDetector) {
      throw new BeagleRuntimeError(
        '"beagleProvider", "ngZone" and "changeDetector" must be set before the AfterViewInit runs. Use the constructor or the component instance to set their values.',
      )
    }
    this.createBeagleView()
    this.createEventHandler()
    this.view.updateWithFetch(this.loadParams)
  }

  async ngOnChanges(changes: SimpleChanges) {
    if (changes && changes.loadParams) {
      if (
        changes.loadParams.previousValue
        && changes.loadParams.previousValue !== changes.loadParams.currentValue
      ) {
        this.view.updateWithFetch(this.loadParams)
      }
    }
  }

  ngOnDestroy() {
    BeagleContext.unregisterView(this.viewId)
  }
}
