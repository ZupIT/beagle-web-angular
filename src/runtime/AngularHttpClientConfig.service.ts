/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

export interface AngularHttpClientConfig {
  setService: (service: any, method: string) => void,
  requestBeagle: (url: RequestInfo, optionsRequest?: RequestInit) => Promise<Response>,
}

function createBeagleHttpClient(): AngularHttpClientConfig {
  let httpService: any
  let requestMethod: string

  return {
    setService: (service, method) => {
      httpService = service
      requestMethod = method
    },
    requestBeagle: (url: RequestInfo, optionsRequest?: RequestInit) => {
      if (!httpService || !requestMethod) {
        throw new Error('Undefined service or method. Please set them before using')
      }

      if (!httpService[requestMethod]) {
        throw new Error('Invalid service method. Please set a valid method')
      }

      const { body, headers, method } = optionsRequest || {}
      const httpRequestOptions = { observe: 'response', body, headers }
  
      return new Promise(async (resolve) => {
        const request =
          await httpService[requestMethod](url, method || 'get', httpRequestOptions).toPromise()
        request.json = async () => request.body
        request.text = async () => {
          try {
            return JSON.stringify(request.body)
          } catch (error) {
            return request.body
          }
        }
        resolve(request)
      })
    },
  }
}

export const AngularHttpClientConfig = createBeagleHttpClient()
