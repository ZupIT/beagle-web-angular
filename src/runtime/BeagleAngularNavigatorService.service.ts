/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/
import { EventEmitter, Injectable, Output } from '@angular/core'
import { BeagleNavigator, BeagleView, DefaultWebNavigator, NavigationType, Route } from '@zup-it/beagle-web'
import { BeagleProvider } from './BeagleProvider.service'
import BeagleRuntimeError from './errors'

export interface ViewWidget {
  id: string,
  view: BeagleView,
}

@Injectable()
export class BeagleAngularNavigatorService {
  @Output() public onChange = new EventEmitter<ViewWidget>()
  
  private navigator: BeagleNavigator<ViewWidget>
  private nextViewId = 1
  
  constructor (private beagleProvider: BeagleProvider) {
  }

  public create(): void {
    const beagleService = this.beagleProvider.getBeagleUIService()
    if (!beagleService) {
      throw new BeagleRuntimeError('You need to start the beagle provider before using a remote view.')
    }

    this.navigator = DefaultWebNavigator.create(beagleService, this.widgetBuilder)
    this.navigator.onChange((widget: ViewWidget) => {
      this.onChange.emit(widget)
    })
  }

  private widgetBuilder = (view: BeagleView): ViewWidget => {
    const widget = {
      id: `${this.nextViewId++}`,
      view: view,
    }
    return widget
  }

  pushStack = (route: Route, controllerId?: string) => this.navigator.pushStack(route, controllerId)
  
  popStack = () => this.navigator.popStack()
  
  pushView = (route: Route) => this.navigator.pushView(route)
  
  popView = () => this.navigator.popView()
  
  popToView = (route: string) => this.navigator.popToView(route)
  
  resetStack = (route: Route, controllerId?: string) => 
    this.navigator.resetStack(route, controllerId)
  
  resetApplication = (route: Route, controllerId?: string) => 
    this.navigator.resetApplication(route, controllerId)

  navigate = (type: NavigationType, route?: Route | string, controllerId?: string) => 
    this.navigator.navigate(type, route, controllerId)
  
  isEmpty = () => this.navigator.isEmpty()
  
  getCurrentRoute = () => this.navigator.getCurrentRoute()
}
