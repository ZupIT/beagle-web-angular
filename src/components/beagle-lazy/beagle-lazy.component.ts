/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import { Component, Input, AfterViewInit } from '@angular/core'
import { LoadParams } from '@zup-it/beagle-web'
import { BeagleComponent } from '../../runtime/BeagleComponent'
import { BeagleLazyInterface } from '../schemas/lazy'

@Component({
  selector: 'beagle-lazy',
  templateUrl: './beagle-lazy.component.html',
})
export class BeagleLazyComponent extends BeagleComponent implements BeagleLazyInterface,
  AfterViewInit {

  @Input() path: string

  constructor() {
    super()
  }

  ngAfterViewInit() {
    const params: LoadParams = {
      path: this.path,
      shouldShowLoading: false,
    }

    this.getViewContentManager().replaceComponent(params)
  }

}
