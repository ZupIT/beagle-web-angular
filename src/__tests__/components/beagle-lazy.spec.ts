/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import { TestBed, async, ComponentFixture, tick, fakeAsync } from '@angular/core/testing'
import { BeagleLazyComponent } from '../../components/beagle-lazy/beagle-lazy.component'
import { BeagleComponent } from '../../runtime/BeagleComponent'
import mockViewContentManager from './mocks/test-mocks.spec'

let component: BeagleLazyComponent
let beagleComponent: BeagleComponent
let fixture: ComponentFixture<BeagleLazyComponent>


describe('BeagleLazyComponent', () => {

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [
        BeagleLazyComponent,
      ],
      providers: [
        BeagleComponent,
      ],
    }).compileComponents()

    fixture = TestBed.createComponent(BeagleLazyComponent)
    beagleComponent = TestBed.get(BeagleComponent)
    component = fixture.componentInstance
    component.path = 'http://test.com.br'
  }))

  it('should match snapshot', () => {
    expect(component).toMatchSnapshot()
  })

})
