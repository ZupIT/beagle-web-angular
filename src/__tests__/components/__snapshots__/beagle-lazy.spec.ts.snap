// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BeagleLazyComponent should match snapshot 1`] = `
BeagleLazyComponent {
  "path": "http://test.com.br",
}
`;
